### PR TITLE
[FLYW-425] Add ability to set providers on groups and projects

### PR DESCRIPTION
**Pull request by ehlertjd**
### Review Checklist

- Tests were added to cover all code changes
- Documentation was added / updated
- Code and tests follow standards in CONTRIBUTING.md
- Extra care is taken and log messages are created for [Critical Data](https://github.com/flywheel-io/core/wiki/Critical-Data-Flows)

### CODE REVIEW

File: api/site/providers/repository.py [hunk 1]
```diff
@@ -131,6 +133,74 @@ def update_provider(provider_id, doc):
...
+    3. A storage provider isn't already set
+
+    Setting either provider to the current value is a no-op and doesn't
+    trigger authorization errors.
+
+    Side-effect: This will convert any IDs in the links parameter to ObjectIds.
+
+    Args:
+        container (dict): The current container (or empty if it doesn't exist)
+        links (dict): The provider links to update.
```
Comment (hkethi002) [inline api/site/providers/repository.py]: I find links to be a confusing word to use here, can't we just call them provider_ids

Comment (ehlertjd) [reply api/site/providers/repository.py]: Agreed, I changed the variable & function names.

File: api/site/providers/repository.py [hunk 2]
```diff
@@ -131,6 +133,74 @@ def update_provider(provider_id, doc):
...
+        is_admin (bool): Whether or not the user is a site administrator
+
+    Raises:
+        APIPermissionException: If the user is unautorized to make the change.
+        APIValidationException: If the user attempted an invalid transition.
+        APINotFoundException: If the given storage provider does not exist.
```
Comment (nagem) [inline api/site/providers/repository.py]: I like the additional details in the function signature, especially the `Raises` 👍

File: api/site/providers/repository.py [hunk 3]
```diff
@@ -131,6 +133,66 @@ def update_provider(provider_id, doc):
...
+
+    for provider_class in ('compute', 'storage'):
+        updates[provider_class] = False
+        if provider_class in provider_ids:
+            # Ensure ObjectId
+            provider_ids[provider_class] = bson.ObjectId(provider_ids[provider_class])
+            current_id = current_provider_ids.get(provider_class)
+            if current_id != provider_ids[provider_class]:
+                if current_id:
+                    raise errors.APIValidationException('Cannot change {} provider once set!'.format(provider_class))
```
Comment (nagem) [inline api/site/providers/repository.py]: If a site admin sets a group provider, does that mean the project provider remains empty and is collected from the group level until one at the project level is set?

Comment (ehlertjd) [reply api/site/providers/repository.py]: @nagem That's a good question. Thinking about it, I think we do have to propagate down from the group level (setting on any project that hasn't already been set.)

If we start using implicit provider selection, then we could end up with data split across projects if you were ever to finally set a different provider on a project that had implicitly been using the group provider.

Comment (ehlertjd) [reply api/site/providers/repository.py]: As discussed offline, we will not propagate changes, and will look up to find the provider.

**Pull request merged**